mlflow/types/utils.py

```diff
@@ -79,7 +79,75 @@ def clean_tensor_type(dtype: np.dtype):
     return dtype
 
 
-def _infer_datatype(data) -> DataType:
+def _infer_colspec_type(data: Any) -> Union[DataType, Array, Object]:
+    """
+    Infer an MLflow Colspec type from the dataset.
+
+    :param data: data to infer from.
+    :return: Object
+    """
+    dtype = _infer_datatype(data)
+    _validate_inferred_type(dtype)
+    return dtype
+
+
+def _infer_datatype(data: Any) -> Union[DataType, Array, Object]:
+    if isinstance(data, dict):
+        properties = [Property(name=k, dtype=_infer_datatype(v)) for k, v in data.items()]
+        return Object(properties=properties)
+
+    if isinstance(data, (list, np.ndarray)):
+        return _infer_array_datatype(data)
+
+    return _infer_scalar_datatype(data)
+
+
+def _infer_array_datatype(data: Union[List, np.ndarray]) -> Optional[Array]:
+    """Infer schema from an array. This tries to infer type if there is at least one
+    non-null item in the list, assuming the list has a homogeneous type. However,
+    if the list is empty or all items are null, returns None as a sign of undetermined.
+
+    E.g.
+        ["a", "b"] => Array(string)
+        ["a", None] => Array(string)
+        [["a", "b"], []] => Array(Array(string))
+        [] => None
+
+    :param data: data to infer from.
+    :return: Array(dtype) or None if undetermined
+    """
+    result = None
+    for item in data:
+        # We accept None in list to provide backward compatibility,
+        # but ignore them for type inference
+        if _is_none_or_nan(item):
+            continue
+
+        dtype = _infer_datatype(item)
+
+        if dtype is not None:
+            if result is None:
+                result = Array(dtype)
```

Comment:
```suggestion
        if dtype is None:
            continue
            
            
        if result is None:
            result = Array(dtype)
            continue
            
        ...
```

I'd consider using more early-continue.

Reply:
I agree with the first one to reduce indent level, tho the latter one is not necessary? I'm personally not big fan of those jumping syntax..., while open to change if there is a common style guide we use?

Reply:
> tho the latter one is not necessary?

Agreed, let's apply the first one